askama_derive/src/generator.rs

```diff
@@ -576,35 +593,70 @@ impl<'a> Generator<'a> {
         self.locals.push();
         self.handle(ctx, nodes, buf, AstLevel::Block);
         self.locals.pop();
-        self.flush_ws(buf, *ws2);
+        self.flush_ws(*ws2);
 
         // Restore original block context and set whitespace suppression for
         // succeeding whitespace according to the outer WS spec
         self.super_block = prev_block;
         self.prepare_ws(outer);
     }
 
-    fn write_expr(&mut self, buf: &mut Buffer, ws: WS, s: &Expr) {
-        self.handle_ws(buf, ws);
-        let mut expr_buf = Buffer::new(0);
-        let wrapped = self.visit_expr(&mut expr_buf, s);
-
-        use self::DisplayWrap::*;
-        use super::input::EscapeMode::*;
-        buf.writeln("write!(writer, \"{}\", &");
-        buf.write(&match (wrapped, &self.input.escaping) {
-            (Wrapped, &Html) | (Wrapped, &None) | (Unwrapped, &None) => expr_buf.buf,
-            (Unwrapped, &Html) => format!("::askama::MarkupDisplay::from(&{})", expr_buf.buf),
-        });
-        buf.writeln("");
-        buf.writeln(")?;");
+    fn write_expr(&mut self, ws: WS, s: &'a Expr<'a>) {
+        self.handle_ws(ws);
+        self.buf_writable.push(Writable::Expr(s));
     }
 
-    // Write literals buffer and empty
-    fn write_buf_lit(&mut self, buf: &mut Buffer) {
-        if !self.buf_lit.is_empty() {
-            buf.writeln(&format!("writer.write_str({:#?})?;", self.buf_lit.join("")));
-            self.buf_lit = vec![];
+    // Write expression buffer and empty
+    fn write_buf_writable(&mut self, buf: &mut Buffer) {
+        if !self.buf_writable.is_empty() {
```

Comment:
Did you miss my comment about returning early here by inverting the `if` condition?

Reply:
I did not understand you. I flipped the if below. You want
```
if self.buf_writable.is_empty() {
   return
```
??
Can you put the code, please?

Reply:
Yes, that is exactly what I want. 👍

Reply:
Done

Reply:
I'll elide the `else` clause here.